MsixCore/msixmgr/msixmgr.cpp

```diff
@@ -173,6 +173,60 @@ void OutputUnpackFailures(
     }
 }
 
+void OutputUnpackFailuresToFile(
+    _In_ std::wstring packageSource,
+    _In_ std::vector<std::wstring> skippedFiles,
+    _In_ std::vector<std::wstring> failedPackages,
+    _In_ std::vector<HRESULT> failedPackagesErrors,
+    _In_ std::wstring outfilePath)
+{
+    std::wofstream  outfile;

```

Comment:
>std::wofstream [](http://example.com/codeflow?start=4&length=14)

Duplicating all this code for writing to the different types of logs should be unnecessary, but writing multiple streams at once also seems to have some complexity. Easiest solution seems like building a big wstringstream of everything that needs to get logged in this function and then having the wrapper function either write that to just std::wcout or both std::wcout and the file stream.

std::wstringstream unpackFailureString;
use unpackFailureString everywhere std::wcout is used.
add unpackFailureString << std::flush;

then do:
    std::wcout << unpackFailureString.rdbuf() << std::endl;
if (!outputFilePath.empty)
{
open
   outFile << unpackFailureString.rdbuf() << std::endl;
}

Reply:
The duplication works because we want to have different behavior when output path is set and when it isn't this is something I've validated with PM during design.

Reply:
Duplicating all the output string logic isn't necessary to have different behavior when output path is set.


void GetUnpackOutputFailures(
    _In_ std::wstring packageSource,
    _In_ std::vector<std::wstring> skippedFiles,
    _In_ std::vector<std::wstring> failedPackages,
    _In_ std::vector<HRESULT> failedPackagesErrors,
    _In_ std::wstringstream& unpackFailureStringStream)
{
    if (!skippedFiles.empty())
    {
        unpackFailureStringStream << std::endl;
        unpackFailureStringStream << "[WARNING] The following items from " << packageSource << " were ignored because they are not packages or bundles " << std::endl;
...
etc
}

void OutputUnpackFailures(
    _In_ std::wstring packageSource,
    _In_ std::vector<std::wstring> skippedFiles,
    _In_ std::vector<std::wstring> failedPackages,
    _In_ std::vector<HRESULT> failedPackagesErrors,
    _In_ std::wstring outputFilePath)
{
    std::wstringstream unpackFailureStringStream;
    GetUnpackOutputFailures(packageSource, skippedFiles, failedPackages, failedPackagesErrors, unpackFailureStringStream);
    std::wstring unpackFailureString = unpackFailureStringStream.str();
    std::wcout << unpackFailureString << std::endl;
    if (!outputFilePath.empty())
    {
        std::wofstream  outFile;
        outFile.open(outputFilePath);
        outFile << unpackFailureString << std::endl;
        outFile.close();
        wcout << "Wrote error output to file: " << outputFilePath;
    }
}

And then just call the OutputUnpackFailures function and let it deal with the if\else for if outputFilePath is set.
std::wcout << std::endl;
std::wcout << "Finished unpacking packages to: " << unpackDestination << std::endl;
std::wcout << std::endl;

OutputUnpackFailures(packageSourcePath, skippedFiles, failedPackages, failedPackagesErrors, outputFilePath);

Reply:
Completed making those changes https://github.com/microsoft/msix-packaging/pull/600/commits/82a4c90d06d399891ba0dcfa67ab71904f1b30e9